crates/accelerate/src/unitary_synthesis.rs

```
@@ -0,0 +1,1094 @@
+                basis_2q_fidelity,
+                basis_1q,
+                None,
+            )?;
+
+            decomposers.push(DecomposerElement {
+                decomposer: DecomposerType::TwoQubitBasisDecomposer(Box::new(decomposer)),
+                decomp_gate: gate.operation.name().to_string(),
+                decomp_gate_params: Some(gate.params.clone()),
+            });
+        }
+    }
+
+    // If our 2q basis gates are a subset of cx, ecr, or cz then we know TwoQubitBasisDecomposer
+    // is an ideal decomposition and there is no need to bother calculating the XX embodiments
+    // or try the XX decomposer
+    let available_basis_set: HashSet<&str> = available_2q_basis.keys().copied().collect();
+    let goodbye_set: HashSet<&str> = vec!["cx", "cz", "ecr"].into_iter().collect();
+    if goodbye_set.is_superset(&available_basis_set) {
+        // TODO: decomposer cache thingy
+        return Ok(Some(decomposers));
+    }
+
+    // Let's now look for possible controlled decomposers (i.e. XXDecomposer)
+    let controlled_basis: IndexMap<&str, NormalOperation> = available_2q_basis
+        .into_iter()
+        .filter(|(_, v)| is_controlled(v))
+        .collect();
+    let mut pi2_basis: Option<&str> = None;
+    let xx_embodiments = PyModule::import_bound(py, "qiskit.synthesis.two_qubit.xx_decompose")?
+        .getattr("XXEmbodiments")?;
+
+    // The xx decomposer args are the interaction strength (f64), basis_2q_fidelity (f64),
+    // and embodiments (Bound<'_, PyAny>).
+    let xx_decomposer_args = controlled_basis.iter().map(
+        |(name, op)| -> PyResult<(f64, f64, pyo3::Bound<'_, pyo3::PyAny>)> {
+            let strength = 2.0
+                * TwoQubitWeylDecomposition::new_inner(
+                    op.operation.matrix(&op.params).unwrap().view(),
+                    None,
+                    None,
+                )
+                .unwrap()
+                .a;
```

Since we return a pyresult from the map here we can return the error from `new_inner()` if there is one:

```suggestion
                )?
                .a;
```